### ISSUE TITLE
Remove std::unary_function and std::binary_function to support C++17

### ISSUE THREAD

**Issue by timstr** [alembic/alembic#203]
I'm trying to compile Alembic using Visual Studio 2017 against the C++17 standard, but I run into these diagnostics: `'unary_function': is not a member of 'std'` and `'binary_function': is not a member of 'std'`. The problem stems from these struct definitions in Alembic/AbcCoreAbstract/ArraySampleKey.h:

```
struct ArraySampleKeyEqualTo :
        public std::binary_function<ArraySampleKey,ArraySampleKey,bool>
               ^^^^^^^^^^^^^^^^^^^^
{
    bool operator()( ArraySampleKey const &a,
                     ArraySampleKey const &b ) const
    {
        return a == b;
    }
};
/* ... */
struct ArraySampleKeyStdHash :
        public std::unary_function<ArraySampleKey,size_t>
               ^^^^^^^^^^^^^^^^^^^
{
    size_t operator()( ArraySampleKey const &a ) const
    {
        return StdHash( a );
    }
};
```
These helper classes were deprecated in C++11 and are removed completely in C++17. A simple fix could be to simply remove them entirely, and manually add the `typedef`s that these classes define, in case anyone depends on those. The following `typedef`s are written according to cppreference.com ([1](https://en.cppreference.com/w/cpp/utility/functional/unary_function),[2](https://en.cppreference.com/w/cpp/utility/functional/binary_function)). I propose this replacement:

```
struct ArraySampleKeyEqualTo
{
    typedef ArraySampleKey first_argument_type;
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    typedef ArraySampleKey second_argument_type;
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    typedef bool result_type;
    ^^^^^^^^^^^^^^^^^^^^^^^^

    bool operator()( ArraySampleKey const &a,
                     ArraySampleKey const &b ) const
    {
        return a == b;
    }
};
/* ... */
struct ArraySampleKeyStdHash
{
    typedef ArraySampleKey argument_type;
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    typedef size_t result_type;
    ^^^^^^^^^^^^^^^^^^^^^^^^^^

    size_t operator()( ArraySampleKey const &a ) const
    {
        return StdHash( a );
    }
};
```
Thanks for considering!

**Comment by lamiller0** [comment]
Thank you for reporting this.

Was this the only C++17 thing that bit you?

**Comment by timstr** [comment]
This was all that was keeping me from compiling, yes. But I haven't attempted to compile every header. I've done some basic searches and I can't find any other mentions of `std::unary_function`, `std::binary_function`, or any of the other function objects that were removed with C++17 (`ptr_fn`, `mem_fun`, and `mem_fun_ref`). I did just find some other things after a bit of research. I'm reading [this document](https://isocpp.org/files/papers/p0636r0.html) to see what's changed.

`std::auto_ptr`, which is also removed in C++17, is used in arnold/Procedural/ProcMain.cpp and in prman/Procedural/ProcMain.cpp. From some quick reading, it seems like these can be replaced with `std::unique_ptr`.

`std::bind2nd` is also removed, and appears in maya/AbcExport/AbcExport.cpp. I believe the relevant snippets can be replaced with lambdas, something like the following:

```
- std::bind2nd(std::plus<double>(), sampleRanges.begin()->startTime)
+ [startTime=sampleRanges.begin()->startTime](double x){ return x + startTime; }
```

Thanks for the quick reply!
